utils/check_copyright.py

```diff
@@ -0,0 +1,219 @@
+            elif state is 1:
+                if MIT_BEGIN_RE.search(line):
+                    state = 2
+                    emit = False
+                elif APACHE2_BEGIN_RE.search(line):
+                    # Assume an Apache license is preceded by a copyright
+                    # notice.  So just emit it like the rest of the file.
+                    state = 9
+            elif state is 2:
+                # Replace the MIT license with Apache 2
+                emit = False
+                if MIT_END_RE.search(line):
+                    state = 9
+                    sys.stdout.write(licensed)
+            if emit:
+                sys.stdout.write(line)
+
+
+def alert_if_no_copyright(glob, comment_prefix):
+    """Prints names of all files missing either a copyright or Apache 2 license.
+
+    Finds all glob-matching files under the current directory and checks if they
+    contain the copyright message and license notice.  Prints the names of all the
+    files that don't meet both criteria.
+
+    Returns the total number of file names printed.
+    """
+    printed_count = 0
+    for file in filtered_descendants(glob):
+        has_copyright = False
+        has_apache2 = False
+        line_num = 0
+        apache_expected_end = 0
+        with open(file) as contents:
+            for line in contents:
+                line_num += 1
+                if COPYRIGHT_RE.search(line):
+                    has_copyright = True
+                if APACHE2_BEGIN_RE.search(line):
+                    apache_expected_end = line_num + LICENSED_LEN
+                if (line_num is apache_expected_end) and APACHE2_END_RE.search(line):
+                    has_apache2 = True
+        if not (has_copyright and has_apache2):
+            message = file
+            if not has_copyright:
+                message += ' has no copyright'
+            if not has_apache2:
+                message += ' has no Apache 2 license notice'
+            print(message)
+            printed_count += 1
+    return printed_count
+
+
+class ArgParser(argparse.ArgumentParser):
+    def __init__(self):
+        super(ArgParser, self).__init__(
+                description=inspect.getdoc(sys.modules[__name__]))
+        self.add_argument('--update', dest='author', action='store',
+                          help='For files missing a copyright notice, insert '
+                               'one for the given author, and add a license '
+                               'notice.  The author must be in the AUTHORS '
+                               'list in the script.')
+
+
+def main():
+    glob_comment_pairs = [('*.h', '//'), ('*.hpp', '//'), ('*.sh', '//'),
+                          ('*.py', '#'), ('*.cpp', '//'),
+                          ('CMakeLists.txt', '#')]
+    argparser = ArgParser()
+    args = argparser.parse_args()
+
+    if args.author:
+        if args.author not in AUTHORS:
+            print('error: --author must be in the AUTHORS list in '
```

Comment:
This actually suggests we are using `--author` as the command line option, while the above listed as `--update`?